File: Preview.py
"""Preview file for the retro IDE color schemes.

Open this file with each scheme to check Python coverage:
comments, docstrings, decorators, f-strings, numbers, types, and errors.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Iterable, Optional

MAX_RETRY: int = 3          # constant in capitals
RATE = 0.075                # float
MASK = 0xFF_00             # hex with separator
PATTERN = re.compile(r"^(?P<code>[A-Z]{3})-(\d+)$")     # regular expression


class Currency(Enum):
    """An enumeration. Members show as constants."""

    EUR = "EUR"
    USD = "USD"


@dataclass(frozen=True, slots=True)
class Money:
    """A value with a currency. The decorator above is an attribute."""

    amount: float
    currency: Currency = Currency.EUR
    tags: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.amount < 0:
            raise ValueError(f"negative amount: {self.amount!r}")

    @property
    def label(self) -> str:
        return f"{self.amount:,.2f} {self.currency.value}"      # format specifier

    def __add__(self, other: "Money") -> "Money":
        assert self.currency is other.currency, "currency mismatch"
        return Money(self.amount + other.amount, self.currency)


def total(items: Iterable[Money], *, discount: Optional[float] = None) -> Money:
    # TODO: support more than one currency
    result = Money(0.0)
    for item in items:
        result += item
    if discount is not None and 0 < discount <= 1:
        result = Money(result.amount * (1 - discount), result.currency)
    return result


async def fetch(url: str, timeout: float = 2.5) -> dict[str, object]:
    import json          # local import
    import urllib.request

    with urllib.request.urlopen(url, timeout=timeout) as response:
        payload: bytes = response.read()
    return json.loads(payload.decode("utf-8"))


try:
    print(total([Money(19.99), Money(5.01)]).label)
except (ValueError, AssertionError) as error:      # exception types
    print("failed:", error, sep=" ")
finally:
    del MASK

lambda_sum = lambda a, b: a + b       # noqa: E731
squares = {n: n ** 2 for n in range(10) if n % 2 == 0}
text = """A triple-quoted string
with a second line."""
raw = r"C:\temp\no-escape"

File: Preview.java
/**
 * Preview file for the retro IDE color schemes.
 *
 * Open this file with each scheme to check Java coverage:
 * Javadoc, annotations, generics, fields, local variables, and enums.
 *
 * @author Custom
 * @see Money#add(Money)
 */
package com.example.retro;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;

@FunctionalInterface
interface Rounder {
    double round(double value);
}

enum Currency {
    EUR("€"),
    USD("$");

    private final String symbol;          // private field

    Currency(String symbol) {
        this.symbol = symbol;
    }

    public String symbol() {
        return this.symbol;
    }
}

public final class Money implements Comparable<Money> {

    public static final int MAX_RETRY = 3;         // static final field
    private static final double RATE = 0.075D;
    private static final long MASK = 0xFF00L;

    private final double amount;
    private final Currency currency;

    public Money(double amount, Currency currency) {
        if (amount < 0.0) {
            throw new IllegalArgumentException("negative amount: " + amount);
        }
        this.amount = amount;
        this.currency = Objects.requireNonNull(currency, "currency");
    }

    /**
     * Adds two amounts of the same currency.
     *
     * @param other the other amount
     * @return the sum
     * @throws IllegalStateException if the currencies do not agree
     */
    public Money add(Money other) {
        // TODO: support more than one currency
        if (this.currency != other.currency) {
            throw new IllegalStateException("currency mismatch");
        }
        return new Money(this.amount + other.amount, this.currency);
    }

    @Override
    public int compareTo(Money other) {
        return Double.compare(this.amount, other.amount);
    }

    @Override
    public String toString() {
        return String.format("%.2f%s", this.amount, this.currency.symbol());
    }

    @Deprecated
    public double legacyValue() {
        return this.amount * (1.0 - RATE);
    }

    public static <T extends Number> double sum(List<T> values, Function<T, Double> map) {
        double total = 0.0;
        for (T value : values) {
            total += map.apply(value);
        }
        return total;
    }

    public static void main(String... args) {
        List<Money> items = new ArrayList<>();
        items.add(new Money(19.99, Currency.EUR));
        items.add(new Money(5.01, Currency.EUR));

        Money total = items.stream()
                .reduce(new Money(0.0, Currency.EUR), Money::add);

        Rounder rounder = value -> Math.round(value * 100.0) / 100.0;
        System.out.println(total + " mask=" + MASK + " retry=" + MAX_RETRY);
        System.out.printf("rounded=%s%n", rounder.round(total.legacyValue()));

        char tab = '\t';
        String text = """
                A text block
                with a second line.""";
        assert !text.isEmpty() : "empty" + tab;
    }
}
